stat/combin/combin.go

```diff
@@ -294,11 +304,40 @@ func Cartesian(lens []int) [][]int {
 		}
 		rows *= v
 	}
-	out := make([][]int, rows)
-	for i := 0; i < rows; i++ {
-		out[i] = SubFor(nil, i, lens)
+	return rows
+}
+
+// NewCartesianGenerator returns a CartesianGenerator for iterating over cartesian products which are generated on the fly.
+func NewCartesianGenerator(lens []int) *CartesianGenerator {
+	return &CartesianGenerator{
+		lens: lens,
+		rows: NumCartesianProducts(lens),
+		idx:  -1,
 	}
-	return out
+}
+
+// CartesianGenerator iterates over a cartesian product set.
+type CartesianGenerator struct {
+	lens []int
+	rows int
+	idx  int
+}
+
+// Next moves to the next product of the cartesian set.
+// It returns true if a product can be generated and false if the generator reached the cartesian sets end.
```

Comment:
```suggestion
// It returns false when there are no cartesian product to iterate over anymore.
```

(just a suggestion from a non-native english speaker.)

Reply:
Thanks for your input.
I am also no native english speaker - I shortened the sentence a bit to focus on the false case, but kept the words I find to be important.